codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -272,6 +275,13 @@ private boolean readClientPrefaceString(ByteBuf in) throws Http2Exception {
             if (bytesRead == 0 || !ByteBufUtil.equals(in, in.readerIndex(),
                                                       clientPrefaceString, clientPrefaceString.readerIndex(),
                                                       bytesRead)) {
+                String http1Request =
```

Comment:
I see in your issue description you elude to future optimizations however I think we can avoid the string conversion if the data is not HTTP/1.x:

```diff
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java
index dfe677d..2709c47 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java
@@ -16,6 +16,7 @@ package io.netty.handler.codec.http2;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelHandlerContext;
@@ -198,6 +199,9 @@ public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http
         }
     }
 
+    private static final ByteBuf HTTP_1_X_PREFIX_BUF = Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(
+            new byte[] {'H', 'T', 'T', 'P', '/', '1', '.'})).asReadOnly();
+
     private final class PrefaceDecoder extends BaseDecoder {
         private ByteBuf clientPrefaceString;
         private boolean prefaceSent;
@@ -275,12 +279,10 @@ public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http
             if (bytesRead == 0 || !ByteBufUtil.equals(in, in.readerIndex(),
                                                       clientPrefaceString, clientPrefaceString.readerIndex(),
                                                       bytesRead)) {
-                String http1Request =
-                    in.toString(in.readerIndex(), Math.min(in.readableBytes(), 1024), CharsetUtil.US_ASCII);
-                int http1Index;
-                if ((http1Index = http1Request.indexOf("HTTP/1.")) != -1) {
+                if (ByteBufUtil.equals(HTTP_1_X_PREFIX_BUF, HTTP_1_X_PREFIX_BUF.readerIndex(),
+                                   in, in.readerIndex(), HTTP_1_X_PREFIX_BUF.readableBytes())) {
                     throw connectionError(PROTOCOL_ERROR, "Unexpected HTTP/1.x request: %s",
-                                                          http1Request.substring(0, http1Index));
+                            in.toString(in.readerIndex(), Math.min(in.readableBytes(), 1024), CharsetUtil.US_ASCII));
                 }
                 String receivedBytes = hexDump(in, in.readerIndex(),
                                                min(in.readableBytes(), clientPrefaceString.readableBytes()));

```